common/dal/src/cache.rs

```diff
@@ -0,0 +1,161 @@
+use common_cache::Cache;
+use common_cache::DefaultHashBuilder;
+use common_cache::LruCache;
+use common_cache::LruDiskCache;
+use common_exception::Result;
+use common_metrics::label_counter;
+use common_metrics::label_counter_with_val;
+
+use crate::DataAccessor;
+
+const CACHE_READ_BYTES_FROM_REMOTE: &str = "cache_read_bytes_from_remote";
+const CACHE_READ_BYTES_FROM_LOCAL: &str = "cache_read_bytes_from_local";
+const CACHE_ACCESS_COUNT: &str = "cache_access_count";
+const CACHE_ACCESS_HIT_COUNT: &str = "cache_access_hit_count";
+
+struct CacheDeferMetrics<'a> {
+    tenant_id: &'a str,
+    cluster_id: &'a str,
+    cache_hit: bool,
+    read_bytes: u64,
+}
+
+impl Drop for CacheDeferMetrics<'_> {
+    fn drop(&mut self) {
+        label_counter(CACHE_ACCESS_COUNT, self.tenant_id, self.cluster_id);
+        if self.cache_hit {
+            label_counter(CACHE_ACCESS_HIT_COUNT, self.tenant_id, self.cluster_id);
+            label_counter_with_val(
+                CACHE_READ_BYTES_FROM_LOCAL,
+                self.read_bytes,
+                self.tenant_id,
+                self.cluster_id,
+            );
+        } else {
+            label_counter_with_val(
+                CACHE_READ_BYTES_FROM_REMOTE,
+                self.read_bytes,
+                self.tenant_id,
+                self.cluster_id,
+            );
+        }
+    }
+}
+
+pub struct DalCacheConfig {
+    pub memory_cache_size_mb: u64,
+    pub disk_cache_size_mb: u64,
+    pub disk_cache_root: String,
+    pub tenant_id: String,
+    pub cluster_id: String,
+}
+
+type MemCache = Arc<RwLock<LruCache<OsString, Vec<u8>, DefaultHashBuilder, BytesMeter>>>;
+
+// TODO maybe distinct segments cache and snapshots cache
+#[derive(Clone, Debug)]
+pub struct DalCache {
+    pub disk_cache: Arc<RwLock<LruDiskCache>>,
+    pub mem_cache: MemCache,
+    tenant_id: String,
+    cluster_id: String,
+}
+
+impl DalCache {
+    pub fn create(conf: DalCacheConfig) -> Result<Self> {
+        let disk_cache = Arc::new(RwLock::new(LruDiskCache::new(
+            conf.disk_cache_root,
+            conf.disk_cache_size_mb * 1024 * 1024,
+        )?));
+        Ok(Self {
+            mem_cache: Arc::new(RwLock::new(LruCache::with_meter(
+                conf.memory_cache_size_mb * 1024 * 1024,
+                BytesMeter,
+            ))),
+            disk_cache,
+            tenant_id: conf.tenant_id,
+            cluster_id: conf.cluster_id,
+        })
+    }
+
+    pub async fn read<K: AsRef<OsStr>>(
+        &self,
+        location: K,
+        da: &dyn DataAccessor,
+    ) -> Result<Vec<u8>> {
+        let location = location.as_ref().to_owned();
+        let mut metrics = CacheDeferMetrics {
+            tenant_id: self.tenant_id.as_str(),
+            cluster_id: self.cluster_id.as_str(),
+            cache_hit: false,
+            read_bytes: 0,
+        };
+
+        // get data from memory cache
+        let mut mem_cache = self.mem_cache.write().await;
```

Comment:
It would be better if we acquire a READ lock here. (and upgrade it to WRITE later if necessary)

Reply:
because it is lru cache, it must be a WRITE lock and it will update inner state when reading.

Reply:
oops, sorry I did not notice that the signature of get is `fn get<Q>(&mut self, k: &Q) -> Option<&V>`